File: swervelib.cpp
#include "swervelib.h"

double radians_to_degrees(double radians) {
    double degrees;
    degrees = radians * 180 / PI;
    return degrees;
}

swervelib::swervelib(double width, double length) {
    A = 0, B = 0, C = 0, D = 0;
    // width = wheelbase, length = track width
    R = sqrt(pow(width, 2) + pow(length, 2));
    _width = width;
    _length = length;

    target_WS1 = 0, target_WS2 = 0, target_WS3 = 0, target_WS4 = 0, MAX_WS = 0;
    target_WA1 = 0, target_WA2 = 0, target_WA3 = 0, target_WA4 = 0;

    whl = new wheel();
}

void swervelib::calcWheelVect(double x,
                              double y,
                              double rudder) {

    A = x - rudder * (_length / R);
    B = x + rudder * (_length / R);
    C = y - rudder * (_width / R);
    D = y + rudder * (_width / R);

    target_WS1 = sqrt(pow(B, 2) + pow(C, 2));
    target_WS2 = sqrt(pow(B, 2) + pow(D, 2));
    target_WS3 = sqrt(pow(A, 2) + pow(D, 2));
    target_WS4 = sqrt(pow(A, 2) + pow(C, 2));
    MAX_WS = std::max({target_WS1, target_WS2, target_WS3, target_WS4});

    if (MAX_WS > 1) {
        target_WS1 /= MAX_WS;
        target_WS2 /= MAX_WS;
        target_WS3 /= MAX_WS;
        target_WS4 /= MAX_WS;
    } else {
        target_WA1 = radians_to_degrees(atan2(B, C));
        target_WA2 = radians_to_degrees(atan2(B, D));
        target_WA3 = radians_to_degrees(atan2(A, D));
        target_WA4 = radians_to_degrees(atan2(A, C));
    }

    this->whl->speed1 = target_WS1;
    this->whl->speed2 = target_WS2;
    this->whl->speed3 = target_WS3;
    this->whl->speed4 = target_WS4;

    this->whl->angle1 = 360 - (target_WA1 + 180);
    this->whl->angle2 = 360 - (target_WA2 + 180);
    this->whl->angle3 = 360 - (target_WA3 + 180);
    this->whl->angle4 = 360 - (target_WA4 + 180);
}

File: swervelib.py
#!/usr/bin/env
import math


class Wheel:
    def __init__(self):
        # front right, front left, rear left, rear right
        self.speed1 = 0
        self.speed2 = 0
        self.speed3 = 0
        self.speed4 = 0
        # front right, front left, rear left, rear right
        self.angle1 = 0
        self.angle2 = 0
        self.angle3 = 0
        self.angle4 = 0


class SwerveLib:
    def __init__(self, width, length):
        # width = wheelbase, length = track width
        self._R = math.sqrt(pow(width, 2) + pow(length, 2))
        self._width = width
        self._length = length
        self._target_WS1 = 0
        self._target_WS2 = 0
        self._target_WS3 = 0
        self._target_WS4 = 0
        self._MAX_WS = 0
        self._target_WA1 = 0
        self._target_WA2 = 0
        self._target_WA3 = 0
        self._target_WA4 = 0
        self._A = 0
        self._B = 0
        self._C = 0
        self._D = 0
        self._R = 0
        self._width = 0
        self._length = 0
        self.whl = Wheel()

    def radians_to_degrees(self, radians):
        degrees = radians * 180 / math.pi
        return degrees

        # _target_WS1 = 0, _target_WS2 = 0, _target_WS3 = 0, _target_WS4 = 0, _MAX_WS = 0
        # _target_WA1 = 0, _target_WA2 = 0, _target_WA3 = 0, _target_WA4 = 0
    def calcWheelVect(self, x, y, rudder):
        self._A = x - rudder * (self._length / self._R)
        self._B = x + rudder * (self._length / self._R)
        self._C = y - rudder * (self._width / self._R)
        self._D = y + rudder * (self._width / self._R)
        # _A = _B
        self._target_WS1 = math.sqrt(pow(self._B, 2) + pow(self._C, 2))
        self._target_WS2 = math.sqrt(pow(self._B, 2) + pow(self._D, 2))
        self._target_WS3 = math.sqrt(pow(self._A, 2) + pow(self._D, 2))
        self._target_WS4 = math.sqrt(pow(self._A, 2) + pow(self._C, 2))
        self._MAX_WS = max(self._target_WS1, self._target_WS2, self._target_WS3, self._target_WS4)

        if self._MAX_WS > 1:
            self._target_WS1 /= self._MAX_WS
            self._target_WS2 /= self._MAX_WS
            self._target_WS3 /= self._MAX_WS
            self._target_WS4 /= self._MAX_WS
        else:
            self._target_WA1 = self.radians_to_degrees(math.atan2(self._B, self._C))
            self._target_WA2 = self.radians_to_degrees(math.atan2(self._B, self._D))
            self._target_WA3 = self.radians_to_degrees(math.atan2(self._A, self._D))
            self._target_WA4 = self.radians_to_degrees(math.atan2(self._A, self._C))
        self.whl.speed1 = self._target_WS1
        self.whl.speed2 = self._target_WS2
        self.whl.speed3 = self._target_WS3
        self.whl.speed4 = self._target_WS4

        self.whl.angle1 = 360 - (self._target_WA1 + 180)
        self.whl.angle2 = 360 - (self._target_WA2 + 180)
        self.whl.angle3 = 360 - (self._target_WA3 + 180)
        self.whl.angle4 = 360 - (self._target_WA4 + 180)
